zigpy/device.py

```diff
@@ -47,117 +54,150 @@ class Device(zigpy.util.LocalLogMixin, zigpy.util.ListenableMixin):
-            self._group_scan_handle.cancel()
-        self._group_scan_handle = asyncio.ensure_future(self.group_membership_scan())
+    def schedule_initialize(self) -> Optional[asyncio.Task]:
+        # Already-initialized devices don't need to be re-initialized
+        if not self.is_partially_initialized:
+            self.debug("Skipping initialization, device is fully initialized")
+            self._application.device_initialized(self)
+            return None
 
-    async def group_membership_scan(self) -> None:
-        """Sync up group membership."""
-        for ep_id, ep in self.endpoints.items():
-            if ep_id:
-                await ep.group_membership_scan()
+        self.debug("Scheduling initialization")
+
+        self.cancel_initialization()
+        self._initialize_task = asyncio.create_task(self._initialize())
+
+        return self._initialize_task
+
+    @property
+    def is_partially_initialized(self):
+        """
+        Some devices were not properly initialized and need to have more information
+        requested after they have been joined.
+        """
+
+        if self.node_desc is None or not self.node_desc.is_valid:
+            return True
+
+        if self.status != Status.ENDPOINTS_INIT:
+            return True
 
-    async def get_node_descriptor(self):
+        return False
+
+    async def get_node_descriptor(self) -> zdo.types.NodeDescriptor:
         self.info("Requesting 'Node Descriptor'")
-        try:
-            status, _, node_desc = await self.zdo.Node_Desc_req(
-                self.nwk, tries=2, delay=0.1
+
+        status, _, node_desc = await self.zdo.Node_Desc_req(
+            self.nwk, tries=2, delay=0.1
+        )
+
+        if status != zdo.types.Status.SUCCESS:
+            raise zigpy.exceptions.InvalidResponse(
+                f"Requesting Node Descriptor failed: {status}"
             )
-            if status == zdo.types.Status.SUCCESS:
-                self.node_desc = node_desc
-                self.info("Node Descriptor: %s", node_desc)
-                return node_desc
-            else:
-                self.warning("Requesting Node Descriptor failed: %s", status)
-        except (asyncio.TimeoutError, zigpy.exceptions.ZigbeeException):
-            self.warning("Requesting Node Descriptor failed", exc_info=True)
+
+        self.node_desc = node_desc
+        self.info("Got Node Descriptor: %s", node_desc)
+
+        return node_desc
 
     async def _initialize(self):
-        if self.status == Status.NEW:
+        """
+        Discover all basic information about a device.
+        Can be called multiple times and should be reentrant.
+        """
+
+        # Some devices are improperly initialized and are missing a node descriptor
+        if self.node_desc is None or not self.node_desc.is_valid:
             await self.get_node_descriptor()
+
+        if self.status == Status.NEW:
+            self.status = Status.NODE_DESC
```

Comment:
What if we do this like:

```suggestion

        if self.status == Status.NEW or self.node_desc is None or not self.node_desc.is_valid:
            await self.get_node_descriptor()
            self.status = Status.NODE_DESC
```